### ISSUE TITLE
Thunderbird version not working in 68.0

### ISSUE THREAD

**Issue by DavidBE-UK** [fnxweb/urllink#14]
The Thunderbird version is disabled and marked as incompatible with Thunderbird 68 (now the version offered for download from the Thunderbird download page, but not yet pushed as an update to existing older versions).

**Comment by fnxweb** [comment]
Damnit.  Despite promises, they've had to rip out loads of the legacy extension support as it's gone from Firefox (somewhat understandable).  I hadn't noticed as they've not advertised it and Fedora has hung onto TB 60.8 where it still works.  It probably broke in TB 63.0.

I need to do some serious homework, but I won't have chance for a couple of weeks.  I need to figure out if I can re-use my Firefox update to WebExtensions on TB yet, or whether I need to go through that painful process again ...

**Comment by DavidBE-UK** [comment]
I'm afraid that 2.6 is still not working in Thunderbird 68.1.1

**Comment by fnxweb** [comment]
I think I have 68.0.1, maybe it broke again. What platform are you on?  Are there any symptoms, any sign of it at all?

**Comment by DavidBE-UK** [comment]
I'm on a Windows 10 version 1903 x64 laptop.  URL link is in the Menu>Addons menu, which opens a preferences window, but it doesn't close on clicking OK.

Actually, I may be confusing myself here so I'd better be specific as I get some emails that are actually HTML but they LOOK like plain text and I can think they are unless I check the source. What I'm seeing is that in a real plain text email, URLs are being turned into clickable links, with an entry to open them in a browser in the context menu, but I think that is a built in Thunderbird function; but in HTML emails, URLs that are just typed, not marked as HTML links in the source, are staying as non-clickable text, and those are the ones URL link should work with, right? Because it isn't doing anything on those, and is not adding anything in the context menu to open the selected text if I select the URL. Other than the presence of URL Link in the Addons menu, there's no sign of it, and no symptoms other than non-presence.

**Comment by fnxweb** [comment]
Yes, sounds like it's borked. Is there anything in the console window about it?  (If it's still there, tools / developer menu / console window).

**Comment by DavidBE-UK** [comment]
With Thunderbird freshly started (after a reboot after a Windows update). and one email opened that contains some URLs in the HTML part that are just text, not marked as URLs in the HTML, and one of those URLs selected, the only mentions of URLlink I can find in the console are just these log lines:

WebExtensions: Loading packed extension from D:\Thunderbird\Profiles\tfvbka2f.default\extensions\{139a120b-c2ea-41d2-bf70-542d9f063dfd}.xpi ext-legacy.js:157
WebExtensions: Loading add-on preferences from  D:\Thunderbird\Profiles\tfvbka2f.default\extensions\{139a120b-c2ea-41d2-bf70-542d9f063dfd}.xpi ext-legacy.js:190
WebExtensions: Firing profile-after-change listeners for {139a120b-c2ea-41d2-bf70-542d9f063dfd} ext-legacy.js:194

​Nothing about it in the Errors or Warnings sections. (I am now curious, though, as to why the error section is full of errors from the Thunderbird calendar!)

**Comment by fnxweb** [comment]
Ta. I'll have a dig over the weekend.

Both FF and TB spew internal errors all the time!

**Comment by fnxweb** [comment]
Right, I think I've reproduced it (_68.1_), so I'll see if I can bug fix it over the weekend.
In the meantime, can you please try manually editing your **prefs.js** file?

In case you don't know where it is, go to the _TB menu button / Help / Troubleshooting Information_, then look for _Profile Directory_ and click _Open Directory_.

Make sure you exit Thunderbird before editing the file, as it's re-written on exit.  Look for the preference **extensions.urllink.forcesubmenu** and set it to **true**.  Save and re-run TB.

That's the setting I happened to have and it works;  for me, changing it to **false** breaks it so I'm guessing you have it set to that (the default).

**Comment by fnxweb** [comment]
Ah, OK, having re-run it myself TB removes prefs. it thinks are defaults, so you probably won't have that line!
Look for the other **urllink** entries and add the following (order/location doesn't matter):

`user_pref("extensions.urllink.forcesubmenu", true);`

Maybe make a backup copy of the file as well, just in case.

**Comment by DavidBE-UK** [comment]
I'm afraid that made no difference. I also noticed that the only urllink lines in prefs.js were:

user_pref("extensions.urllink.firsttime", true);
user_pref("extensions.urllink.lastversion", "2.05");

(to which I added your line from above), whereas an old prefs.js had

user_pref("extensions.urllink.firsttime", true);
user_pref("extensions.urllink.lastversion", "2.05");
user_pref("extensions.urllink.sandr.0", "^//||file:///");
user_pref("extensions.urllink.sandr.1", "^([A-Za-z]:)||file:///$1");
user_pref("extensions.urllink.submenu.0", "&www.*");
user_pref("extensions.urllink.submenu.1", "www.*.&com");
user_pref("extensions.urllink.submenu.2", "www.*.&org");
user_pref("extensions.urllink.submenu.3", "www.*.&net");
user_pref("extensions.urllink.submenu.4", "&ftp.*");
user_pref("extensions.urllink.submenu.5", "--");
user_pref("extensions.urllink.submenu.6", "In &Google|http://www.google.com/search?q=*&source-id=Mozilla%20Firefox&start=0");
user_pref("extensions.urllink.submenu.7", "In Wi&kipedia|http://en.wikipedia.org/wiki/Special:Search?search=*&sourceid=mozilla-search");

so I tried adding all those other lines too, but it still made no difference.

I also noticed that in the urllink preferences, clicking to switch to the Edit Search and Replace Strings doesn't actually do anything (as well as the OK button not closing it).

**Comment by fnxweb** [comment]
That's a pain.  OK, well, let me see if I can get to the bottom of the glitch I _can_ see, and we can see if that happens to fix your problem.
I'll also try it on my Windows laptop in case there's a difference there (I'm on Linux).

**Comment by fnxweb** [comment]
Having had another play, I'm actually not able to make it stop working at all;  I mean, it's not _behaving_, and won't till I get to re-write it for WebExtensions.  I can't make any changes, including on a fresh profile, that stop the “Open selection” option appear.

I've also tried to replicate your specific issue in case that's a separate big in and of itself.  I've sent myself an HTML email that uses a fix-width font with various flavours of URL typed, but not made links.  However, TB makes the URLs links before sending, so I've not managed to recreate the situation yet.  I'll try to construct an email and send it via raw SMTP.

**Comment by fnxweb** [comment]
Nope, I've now got an HTML email with un-linked URLs, and TB/URL Link still behaves as I'd expect.  Text already looking like a URL is handled by TB (www.fnxweb.com
 — “Open Selected URL”), and text not quite there gives the URL Link sub-menu (fnxweb.com — “Open Selection >”).

**Comment by DavidBE-UK** [comment]
Well, TB 68.2 it's still not working for me. I'm simply not getting an URL Link menu/sub-menu in the context menu at all. So e.g. with the email I received of your last post above, the www.fnxweb.com is a clickable link in both the HTML view (it's marked as an HTML link in the raw email) and plain text view (which is TB itself turning it into a clickable link), while the fnxweb.com is not a clickable link in either HTML or plain text view, and there's no URL Link 'Open Selection or anything else in the page context menu with that text selected, or in any of the Page Window menus either. Oh, and clicking OK in the URL Link options window still appears to do nothing and still doesn't close the window.

Hmn. I have messages display in a separate window, not as a pane in the main TB window. If you have messages in a pane, I wonder if that's relevant ...

**Comment by DavidBE-UK** [comment]
Aha. That IS the difference. Switching to having the messages in a pane in the main TB Window, I DO get the 'Open Selection' in the context menu, and using it on that fnxweb,com it causes your website to appear in Firefox whether in the HTML or plain-text view of the email. Going back to having the messages in a separate window again (Hamburger>Options>Layout Message Pane NOT ticked) and I lose the Open Selection sub-menu entry again.

So, URL Link works when the user has messages displaying in a pane in the TB window, but doesn't work when the user has messages displaying in a separate window - it simply does not appear in the context menu in that case. And I've been using email clients with messages displaying in a separate window since the mid 1990s, and I'm not changing now - NOT being able to display messages in a separate window is something that means I immediately reject an email client as one I might use.

Don't know if it's fixable ... if it isn't, then I'll just have to give up on URL Link. Which would be sad, but it'll be that rather than me having the messages in a pane in the main TB window.

**Comment by fnxweb** [comment]
Ah, ok, that's probably an overlay not working, then. Each window's different, and they've changed them. I'll see if I can find the the new one for separate windows.

FWIW, I've uploaded a new version that should fix the prefs window.

**Comment by DavidBE-UK** [comment]
The update finally became available last night, so I thought I should confirm that it does fix the prefs window.

BTW, the URL Link add-on page says 'Note that as of Thunderbird 60, the preferences dialogue is only available via the Tools / Add-on Preferences main menu.', but actually the prefs dialogue can also be opened from Hamburger>Add-ons, where it is one of the listed Add-on Options.

**Comment by fnxweb** [comment]
Oh yeah, I noticed that too. Ta, I'll update the page. Hoping to find the page issue at the weekend (we're selling our house ATM so I've little spare time).

**Comment by fnxweb** [comment]
OK, think I've fixed the separate window issue, and have pushed 2.6.2 with that fix.  Seems to take a few days to get approved (maybe still manually reviewing legacy extensions, or TB extensions).  You can get the XPI and manually install and test from file via https://github.com/fnxweb/urllink/tree/thunderbird (you may also be able to view pending versions on the addons site).

**Comment by DavidBE-UK** [comment]
Right. Downloaded from GitHub and manually installed - and, yes, it's fixed and working. Great, thanks.

One note in case this happens to others. After installing and restarting TB, on first try it did work but the Open Selection sub-menu only had 'Unaltered' - all the other entries had disappeared. And in Add-ons URL Link preferences the Edit Menu box with the sub-menu entries had also gone blank. On third press of the 'default' button it re-populated, as then did the Open Selection sub-menu. That _might_ be something to do with something I did in those tests and fiddling I did e.g. in about:preferences for posts earlier in this thread. But I thought I'd warn you in case it happens to everyone and you start getting support requests (no idea why it took 3 presses of the default button to repopulate the list). Anyway, thanks again both for producing the add-on in the first place and for continuing to support it. It must be infuriating when your add-ons are broken by TB and Mozilla changing things in the main programs.

**Comment by fnxweb** [comment]
OK, that's good.

I'm not going to get excited now about issues in the “legacy” code, the new WebExtensions version (done for Firefox) is significantly different.  At this stage, if the current TB version can be coaxed into life in _any_ anyway, I'm happy.
After I've moved house, I hope to look into getting it updated so it's all the same codebase again (just raised issue #17).

Thanks for hanging on!

Fixed by commit d2714d9d8c6e79f1461de39871e91a9d56608bd3

**Comment by DavidBE-UK** [comment]
Ah. There is actually a lingering problem with the preferences. The problem I said about it being blank wasn't a one off as I assumed. Whenever I re-open the preferences the list of menu items is blank again, as is the list of search and replace of search and replace items - and the context menu Open Selection sub-menu is back to only containing 'Unaltered'. Clicking the default button re-populates the lists again, and the Open Selection sub-menu then gets the other options back again - although one loses any extra items in those lists again, of course.

**Comment by DavidBE-UK** [comment]
Sorry, I should have said the preferences problem and the lost context menu sub-menu items recurs after  closing TB then restarting it.

**Comment by fnxweb** [comment]
I can't reproduce any problem;  I've created a new test profile, and freshly installed URL Link.  It works out of the box, and the preference window always opens for me.

Are there any new errors in the dev. console window?

**Comment by DavidBE-UK** [comment]
Can't see anything mentioning URL Link in the console window.

The preferences window always opens for me; the problem is that it keeps opening opens with the list of menu items and the list of search and replace items both blank, so the only submenu item in the context menu when I have an URL or part URL selected is 'Unaltered'. I can repopulate those lists with the default button and get the default submenu items back, and I can add items, but I end up losing them all again. I can see it in the prefs.js file, when I have that opened in a text editor that tells me when it's updated, with it being updated with all the urllink.submenu lines and urllink.sandr lines resetting to "".

I'm not quite sure what the full condition is for those submenu and sandr items being reset to blank, but one repeatable sequence that causes it for me is:

I have TB open with the urlink submenu and sandr items at default.
I close TB.
I re-open TB.
I open the URLLink preferences window (where the submenus and sandr list items are still showing the default values).
I click OK to close the URLLink preferences without making any changes.
At this precise moment (I'm notified of the change in prefs.js by my text editor) the submenu and search-and-replace lines are all set to blank, trying to use URLLink will show a context menu submenu with only 'unaltered', and re-opening the preferences, both lists are blank again.

So, basically the problem is that the submenu and search-and-replace settings are being reset to blank at some point when they shouldn't be and I haven't changed them. I just haven't quite managed to figure out exactly what the circumstances are that trigger it.

**Comment by fnxweb** [comment]
OK, I'll try to walk though that chain of events and add some debug to try to determine what on Earth is going on.

**Comment by fnxweb** [comment]
Soz, been a while;  we moved house just after my last comment then immediately hit COVID which screwed everything.  Finally catching up with some of this.

I can't reproduce this with the latest version of the extension (what I have locally is slightly advanced on what's uploaded) and the latest TB I have (91.9).  I see they've migrated extension settings into a DB so it's not in prefs.js any more, maybe that's improved things?

**Comment by DavidBE-UK** [comment]
Ah ... oh ... um ... Thing is, I'm disabled in a way that leaves me in constant pain, and am permanently on prescription painkillers that completely fog my head and have a severe effect on my memory. I've just re-read this thread, and I don't have even the faintest recollection of writing those posts, or whatever train of thought and investigation I was following.

All I can say is, from TB 91 I moved to the betas when they were 92 - currently I'm on TB 101.0b4, expecting 102 today or tomorrow. And I've been using URL Link the whole time. I've been on 3.5.2 since it came out. What I have been doing is changing the strict_max_version in the manifest file each time the TB beta major version changes and reinstalling the extension,  so currently on 3.5.2 but with the strict_max_version in the manifest file changed to 101.*. I am not conscious of needing to make use of URL Link much, and in the maybe once every few weeks i do it's always selecting a non-link URL in an email, usually no http:// and often no www., and going to the URL Link submenu in the context menu and clicking on Unaltered. The submenu is always there when something is selected, fully populated, and at least with Unaltered and a selected URL that has an URL 'end' - .com, .co.uk etc - it has always been causing the expected page to open in Firefox (my default browser).

I can't attest that any sub-menu option other than Unaltered works, or whether changing things in the settings works or not, but in my limited use case 3.5.2, with the strict_max_version changed, has worked consistently in the TB betas up to and including 101.0b4. So from my POV you can close this issue (which, as I say, with my prescription-drug-side-effect futzed brain i don't even remember opening). Thanks.

**Comment by fnxweb** [comment]
Ok, not a problem and thanks for the feedback.

Hmm, I used to have to log in and update the version compatibility fields but stopped as my local copy started carrying on working automatically. You'll have to let me know if you do have to do that and I'll try to get into the habit of updating the entry.

So request any other issues you have, I'll try to not take 2 years to investigate them in the future!
